ak.num on typetracer array raising AssertionError (attempt to set values of TypeTracerArray)

### Version of Awkward Array

1.8.0rc6

### Description and code to reproduce

Trying to call `ak.num` with `axis=0` on a  typetracer array raises an exception

Minimal reproducer (with `import awkward._v2 as ak`):

```python
In [28]: x = ak.Array([[[1, 2, 3], [], [4]], [[3], []]])

In [29]: x
Out[29]: <Array [[[1, 2, 3], [], [4]], [[3], []]] type='2 * var * var * int64'>

In [31]: ak.Array(x.layout.typetracer)
Out[31]: <Array-typetracer type='?? * var * var * int64'>

In [32]: x_tt = ak.Array(x.layout.typetracer)

In [33]: ak.num(x_tt, axis=1)
Out[33]: <Array-typetracer type='?? * int64'>

In [34]: ak.num(x_tt, axis=2)
Out[34]: <Array-typetracer type='?? * var * int64'>

In [35]: ak.num(x_tt, axis=0)
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [35], in <cell line: 1>()
----> 1 ak.num(x_tt, axis=0)

File ~/.pyenv/versions/3.10.2/envs/dask-awkward/lib/python3.10/site-packages/awkward/_v2/operations/structure/ak_num.py:72, in num(array, axis, highlevel, behavior)
      9 """
     10 Args:
     11     array: Data containing nested lists to count.
   (...)
     67     <Array [[1.1, 2.2, 3.3], None, [7.7]] type='3 * option[var * float64]'>
     68 """
     69 layout = ak._v2.operations.convert.to_layout(
     70     array, allow_record=False, allow_other=False
     71 )
---> 72 out = layout.num(axis=axis)
     73 if isinstance(out, (ak._v2.contents.Content, ak._v2.record.Record)):
     74     return ak._v2._util.wrap(out, behavior, highlevel)

File ~/.pyenv/versions/3.10.2/envs/dask-awkward/lib/python3.10/site-packages/awkward/_v2/contents/listoffsetarray.py:589, in ListOffsetArray.num(self, axis, depth)
    587 if posaxis == depth:
    588     out = ak._v2.index.Index64.empty(1, self._nplike)
--> 589     out[0] = self.length
    590     return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
    591         0
    592     ]
    593 elif posaxis == depth + 1:

File ~/.pyenv/versions/3.10.2/envs/dask-awkward/lib/python3.10/site-packages/awkward/_v2/index.py:148, in Index.__setitem__(self, where, what)
    147 def __setitem__(self, where, what):
--> 148     self._data[where] = what

File ~/.pyenv/versions/3.10.2/envs/dask-awkward/lib/python3.10/site-packages/awkward/_v2/_typetracer.py:281, in TypeTracerArray.__setitem__(self, where, what)
    280 def __setitem__(self, where, what):
--> 281     raise AssertionError(
    282         "bug in Awkward Array: attempt to set values of a TypeTracerArray"
    283     )

AssertionError: bug in Awkward Array: attempt to set values of a TypeTracerArray

In [36]: ak.num(x, axis=0)
Out[36]: 2
```

--- a/src/awkward/_v2/contents/bytemaskedarray.py
+++ b/src/awkward/_v2/contents/bytemaskedarray.py
@@ -482,11 +482,11 @@ class ByteMaskedArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         else:
             numnull = ak._v2.index.Index64.empty(1, self._nplike, dtype=np.int64)
             nextcarry, outindex = self._nextcarry_outindex(numnull)

--- a/src/awkward/_v2/contents/emptyarray.py
+++ b/src/awkward/_v2/contents/emptyarray.py
@@ -145,11 +145,11 @@ class EmptyArray(Content):
         posaxis = self.axis_wrap_if_negative(axis)
 
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         else:
             out = ak._v2.index.Index64.empty(0, self._nplike)
             return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)

--- a/src/awkward/_v2/contents/indexedarray.py
+++ b/src/awkward/_v2/contents/indexedarray.py
@@ -416,11 +416,11 @@ class IndexedArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         else:
             return self.project().num(posaxis, depth)
 

--- a/src/awkward/_v2/contents/indexedoptionarray.py
+++ b/src/awkward/_v2/contents/indexedoptionarray.py
@@ -480,11 +480,11 @@ class IndexedOptionArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         _, nextcarry, outindex = self._nextcarry_outindex(self._nplike)
         next = self._content._carry(nextcarry, False, NestedIndexError)
         out = next.num(posaxis, depth)

--- a/src/awkward/_v2/contents/listarray.py
+++ b/src/awkward/_v2/contents/listarray.py
@@ -840,11 +840,11 @@ class ListArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         elif posaxis == depth + 1:
             tonum = ak._v2.index.Index64.empty(self.length, self._nplike)
             assert (

--- a/src/awkward/_v2/contents/listoffsetarray.py
+++ b/src/awkward/_v2/contents/listoffsetarray.py
@@ -593,11 +593,11 @@ class ListOffsetArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         elif posaxis == depth + 1:
             tonum = ak._v2.index.Index64.empty(self.length, self._nplike)
             assert (

--- a/src/awkward/_v2/contents/numpyarray.py
+++ b/src/awkward/_v2/contents/numpyarray.py
@@ -354,11 +354,11 @@ class NumpyArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         shape = []
         reps = 1
         size = self.length

--- a/src/awkward/_v2/contents/recordarray.py
+++ b/src/awkward/_v2/contents/recordarray.py
@@ -430,13 +430,12 @@ class RecordArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            single = ak._v2.index.Index64.empty(1, self._nplike)
-            single[0] = self.length
+            npsingle = self._nplike.full((1,), self.length, np.int64)
+            single = ak._v2.index.Index64(npsingle, nplike=self._nplike)
             singleton = ak._v2.contents.numpyarray.NumpyArray(
                 single, None, None, self._nplike
             )
             contents = [singleton] * len(self._contents)
-
             record = ak._v2.contents.recordarray.RecordArray(
                 contents, self._fields, 1, None, self._parameters, self._nplike
             )

--- a/src/awkward/_v2/contents/regulararray.py
+++ b/src/awkward/_v2/contents/regulararray.py
@@ -606,11 +606,11 @@ class RegularArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         elif posaxis == depth + 1:
             tonum = ak._v2.index.Index64.empty(self.length, self._nplike)
             assert tonum.nplike is self._nplike

--- a/src/awkward/_v2/contents/unionarray.py
+++ b/src/awkward/_v2/contents/unionarray.py
@@ -631,11 +631,11 @@ class UnionArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self.nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         else:
             contents = []
             for content in self._contents:

--- a/src/awkward/_v2/contents/unmaskedarray.py
+++ b/src/awkward/_v2/contents/unmaskedarray.py
@@ -228,11 +228,11 @@ class UnmaskedArray(Content):
     def num(self, axis, depth=0):
         posaxis = self.axis_wrap_if_negative(axis)
         if posaxis == depth:
-            out = ak._v2.index.Index64.empty(1, self._nplike)
-            out[0] = self.length
-            return ak._v2.contents.numpyarray.NumpyArray(out, None, None, self._nplike)[
-                0
-            ]
+            out = self.length
+            if ak._v2._util.isint(out):
+                return np.int64(out)
+            else:
+                return out
         else:
             return ak._v2.contents.unmaskedarray.UnmaskedArray(
                 self._content.num(posaxis, depth), None, self._parameters, self._nplike

